snapcraft/plugins/_ros/ros2.py

```diff
@@ -0,0 +1,181 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2017 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import contextlib
+import os
+import logging
+import shutil
+import subprocess
+
+from snapcraft import (
+    repo,
+    sources,
+)
+
+logger = logging.getLogger(__name__)
+
+_ROS2_URL_TEMPLATE = (
+    'https://raw.githubusercontent.com/ros2/ros2/{version}/ros2.repos'
+)
+
+
+_INSTALL_TOOLS_STEP = 'install-tools'
+_FETCH_ROS2_STEP = 'fetch-ros2'
+_BUILD_ROS2_STEP = 'build-ros2'
+
+
+class Bootstrapper:
+    """Bootstrap ROS2 by building the underlay from source."""
+
+    def __init__(self, *, version, bootstrap_path, ubuntu_sources, project):
+        """Initialize bootstrapper.
+
+        :param str version: The ROS2 version to boostrap (e.g. release-beta3)
+        :param str bootstrap_path: Working directory for the bootstrap process
+        :param str ubuntu_sources: Ubuntu repositories from which dependencies
+                                   will be fetched.
+        :param project: Instance of ProjectOptions for project-wide settings
+        :type project: snapcraft.ProjectOptions
```

Comment:
why not call it project_options?

Reply:
I can, although we're not consistent on this. It's called `project` in the plugin, as well as rosdep, and I'm sure elsewhere. `project` is shorter and has the same meaning to me; do you feel that `project_options` is clearer?

Reply:
I do, because the object is not Project, it's ProjectOptions. Maybe file a bitesize bug for later to make it consistent in all plugins?

Reply:
Leave it as project and rename the class (with backwards compat).

Reply:
Alright, I left it.